src/metatrain/experimental/soap_bpnn/spherical.py

```diff
@@ -0,0 +1,401 @@
+                    ],
+                    self.cgs[
+                        str(lam)
+                        + "_"
+                        + str(self.o3_lambda - lam)
+                        + "_"
+                        + str(self.o3_lambda)
+                    ],
+                )
+            for lam in range(self.o3_lambda):
+                basis[:, :, self.o3_lambda + 1 + lam] = cg_combine(
+                    cg_combine(
+                        sh_1[:, lam * lam : (lam + 1) * (lam + 1)],
+                        sh_2[
+                            :,
+                            (self.o3_lambda - lam - 1)
+                            * (self.o3_lambda - lam - 1) : (
+                                (self.o3_lambda - lam - 1) + 1
+                            )
+                            * ((self.o3_lambda - lam - 1) + 1),
+                        ],
+                        self.cgs[
+                            str(lam)
+                            + "_"
+                            + str(self.o3_lambda - lam - 1)
+                            + "_"
+                            + str(self.o3_lambda - 1)
+                        ],
+                    ),
+                    vector_3_spherical,
+                    self.cgs[str(self.o3_lambda - 1) + "_1_" + str(self.o3_lambda)],
+                )
+
+        if self.o3_sigma == -1:
+            # multiply by pseudotensor
+            vector_basis_pseudotensor = self.vector_basis_pseudotensor(
+                systems, selected_atoms
+            )
+            vector_1_spherical = vector_basis_pseudotensor[:, :, 0]
+            vector_2_spherical = vector_basis_pseudotensor[:, :, 1]
+            vector_3_spherical = vector_basis_pseudotensor[:, :, 2]
+            vector_1_spherical = vector_1_spherical / torch.sqrt(
+                torch.sum(torch.square(vector_1_spherical), dim=-1, keepdim=True)
+            )
+            vector_2_spherical = vector_2_spherical / torch.sqrt(
+                torch.sum(torch.square(vector_2_spherical), dim=-1, keepdim=True)
+            )
+            vector_3_spherical = vector_3_spherical / torch.sqrt(
+                torch.sum(torch.square(vector_3_spherical), dim=-1, keepdim=True)
+            )
+            pseudoscalar = cg_combine(
+                cg_combine(vector_1_spherical, vector_2_spherical, self.cgs["1_1_1"]),
+                vector_3_spherical,
+                self.cgs["1_1_0"],
+            )
+            basis = basis * pseudoscalar.unsqueeze(1)
+
+        return basis  # [n_atoms, 2*o3_lambda+1, 2*o3_lambda+1]
+
+
+def cg_combine(A, B, C):
+    return torch.einsum("im, in, mnp-> ip", A, B, C)
+
+
+def get_cg_coefficients(l_max):
+    cg_object = ClebschGordanReal()
+    for l1 in range(l_max + 1):
+        for l2 in range(l_max + 1):
+            for L in range(abs(l1 - l2), min(l1 + l2, l_max) + 1):
+                cg_object._add(l1, l2, L)
+    return cg_object
+
+
+class ClebschGordanReal:
+
```

Comment:
I don't really like that we have one more copy of this code … Could you use the CGTensorProduct from rascaline instead? If not, what's missing there? 

This is likely the code that will end up in metatomic (soon™), so it would be nice to make sure it works everywhere we need it.